skt/kernelbuilder.py

```diff
@@ -338,16 +329,71 @@ def mktgz(self, timeout=60 * 60 * 12):
 
         return fpath
 
-    @staticmethod
-    def log2stdout(lines):
+    def run_multipipe(self, args, env=os.environ.copy()):
         """
-        Show `lines` on stdout.
+        Run a process while writing to stdout/file simultaneously.
 
         Args:
-            lines:      list of strings.
+            args:   A command to run in list format.
+            env:    An option environment to set for the subcommand. The
+                    existing environment is used if one is not specified.
+
+        Returns:
+            The return code from the Popen call.
+
         """
-        sys.stdout.write(''.join(lines))
-        sys.stdout.flush()
+        # Set up a basic logger.
+        root = logging.getLogger('multipipe')
+
+        # Don't adjust the parent logger's settings with the changes we make
+        # here.
+        root.propagate = False
+
+        # Set up a handler that writes directly to the build log.
+        file_handler = logging.FileHandler(self.buildlog)
+        file_handler.setLevel(logging.DEBUG)
+
+        # Set up another handler that writes to stdout.
+        stdout_handler = logging.StreamHandler(sys.stdout)
+        stdout_handler.setLevel(logging.DEBUG)
+
+        # Ensure that the log format is unaltered -- just messages are
+        # written without any time prefixes.
+        formatter = logging.Formatter('%(message)s')
+        file_handler.setFormatter(formatter)
+        stdout_handler.setFormatter(formatter)
+
+        # Add the file and stdout handlers to our logger.
+        root.addHandler(file_handler)
+        root.addHandler(stdout_handler)
+
+        # Run the command.
+        logging.debug("Running multipipe command: %s" % ' '.join(args))
+        root.info("$ %s" % ' '.join(args))
+        process = subprocess.Popen(
+            args,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            bufsize=1
+        )
+
+        # Loop over the buffered stdout/stderr from the command and write the
+        # output to our logger.
+        for line in iter(process.stdout.readline, ''):
+            root.info(line.strip())
+
+        # Ensure the process has exited.
+        process.wait()
```

Comment:
Here you should collect the exit code, otherwise you'll get the cached one.
```python
exit_code = process.wait()
```
then return this variable at the bottom of the function.

Reply:
Fixed.